### ISSUE TITLE
[semantic highlighting] Improving results

### ISSUE THREAD

**Issue by qchateau** [clangd/clangd#623]
I'm quite new to clangd development and I have a few question regarding semantic highlighting, there are some things that seem like bug/unimplemented features to me, but maybe there is a reason that I'm not aware of.

1. "auto" storing lambdas are not tokenized (and a comment says that we don't want to mark them as classes). IMO we should mark these as "class" because 1. they are pretty much simplified function objects 2. it's better than nothing. Any opinion ?
2. There are plenty of cases where "decltype" and "auto" are not tokenized (references, arrays, pointers, primitive types), I'm guessing these cases are bugs.
3. Primitive types are not tokenized, is there any reason for this ?

I've been working on [this branch](https://github.com/qchateau/llvm-project/tree/semantic-highlight) to make semantic highlighting work the way I'd like it to work. Basically, every time a type appears it is properly tokenized and both 'auto' and 'decltype' keywords are tokenized as what they are deduced to (removing pointers, references and arrays).

I'm thinking about submitting the content of this branch for review but I'd like to have a bit of feedback first. I'm guessing a GitHub issue is the right place for this ?

**Comment by HighCommander4** [comment]
>    3. Primitive types are not tokenized, is there any reason for this ?

This used to happen, but was [removed](https://reviews.llvm.org/D69431), because it tended to duplicate client-side highlighting of keywords that clients were already doing.

We don't highlight other keywords, literal tokens (e.g. string literals, character literlas, numerical literals), and comments for the same reason -- clients can (and generally do) highlight these fairly easily without requiring a semantic understanding of the code.

**Comment by HighCommander4** [comment]
>    1. "auto" storing lambdas are not tokenized (and a comment says that we don't want to mark them as classes). IMO we should mark these as "class" because 1. they are pretty much simplified function objects 2. it's better than nothing. Any opinion ?

The comment in question is [here](https://searchfox.org/llvm/rev/f6e885ad2a94357f6f4d18ddf26a8111b3df7ed3/clang-tools-extra/clangd/SemanticHighlighting.cpp#69) and was added in [this patch](https://reviews.llvm.org/D67341). I'm not sure about the reason, maybe @ilya-biryukov or @hokein know.

**Comment by HighCommander4** [comment]
>    2. There are plenty of cases where "decltype" and "auto" are not tokenized (references, arrays, pointers, primitive types), I'm guessing these cases are bugs.

For primitive types, [existing testcases](https://searchfox.org/llvm/rev/f6e885ad2a94357f6f4d18ddf26a8111b3df7ed3/clang-tools-extra/clangd/unittests/SemanticHighlightingTests.cpp#387-388) suggest that it is intended to work, so if there are cases that do not work, I think those are bugs, yes.

For references, arrays, and pointers, it sounds like you're suggesting that they be highlighted according to the type after removing the pointer, reference, or array. It's not clear to me that this is necessarily a good thing. The distinction between "pointer to a class type" and "class type" can be an important one (if e.g. you need to know how to use the return value of a function), and it may be more important for the highlighting to differ between those cases than between "pointer to class type" and "pointer to primitive type".

Note that you can revise an `auto` declaration to have the form `auto*`, `auto&`, etc., and then the `auto` should get the inner type's highlighting (there's even a clang-tidy check that suggests revising `auto` declarations in this way).

**Comment by qchateau** [comment]
Thanks for your time, that's very instructive.

> This used to happen, but was [removed](https://reviews.llvm.org/D69431), because it tended to duplicate client-side highlighting of keywords that clients were already doing.

What is the problem with this duplication ? I find it strange from the client POV that classes and typedefs of primitive types are recognized by semantic highlighting, but not the primitive types themselves. It makes consistent highlighting client-side harder.

> For references, arrays, and pointers, it sounds like you're suggesting that they be highlighted according to the type after removing the pointer, reference, or array. It's not clear to me that this is necessarily a good thing. The distinction between "pointer to a class type" and "class type" can be an important one (if e.g. you need to know how to use the return value of a function), and it may be more important for the highlighting to differ between those cases than between "pointer to class type" and "pointer to primitive type".

I agree with you, but that kind of nuance does not seem supported by the protocol. This means that 
- `auto ptr = static_cast<int*>(nullptr)` is not tokenized
- `auto value = MyClass()` is tokenized as "class"

More generally, `auto` (and `decltype`) is sometimes colored as a keyword, sometimes as a type. That's the inconsistency that is bothering me.

My branch tries to make it more consistent by always tokenizing `auto` as a type (primitive or class), but another solution would be to always ignore it (and the client would most likely always color is as a keyword).

**Comment by HighCommander4** [comment]
> > This used to happen, but was [removed](https://reviews.llvm.org/D69431), because it tended to duplicate client-side highlighting of keywords that clients were already doing.
> 
> What is the problem with this duplication ? I find it strange from the client POV that classes and typedefs of primitive types are recognized by semantic highlighting, but not the primitive types themselves. It makes consistent highlighting client-side harder.

Other than more protocol traffic, I can't really think of a problem. Perhaps @ilya-biryukov can provide more background on the motivation for that patch.

> My branch tries to make it more consistent by always tokenizing `auto` as a type (primitive or class)

I think that would be fine. What I was trying to make an argument for is that, in the absence of a highlighting kind specific to pointer types, it's more useful to highlighting "pointer to class" as "primitive" than as "class".

**Comment by qchateau** [comment]
> > My branch tries to make it more consistent by always tokenizing `auto` as a type (primitive or class)
> 
> I think that would be fine. What I was trying to make an argument for is that, in the absence of a highlighting kind specific to pointer types, it could be more useful to highlight "pointer to class" as "primitive" than as "class".

It feels more natural to me the other way around but that's a point I totally understand. In any case I highly prefer classifying "pointer to class" as "primitive" than nothing at all. 
How would you classify references ? (`decltype(auto)` is often a reference)

**Comment by HighCommander4** [comment]
> In any case I highly prefer classifying "pointer to class" as "primitive" than nothing at all.

Makes sense. I think the reason people haven't noticed things like this is that default themes tend to use the same color for keywords and primitives.

> How would you classify references ? (`decltype(auto)` is often a reference)

I could go either way on that, as references don't affect the syntax of how the object is used, but they can have lifetime implications.

To be clear, all of this is my opinion; I'm not the one who would make a decision about a change in this area.
